File: Hash_Table/problem41/problem41.cpp
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        const int n = static_cast<int>(nums.size());
        for (int i = 0; i < n; ++i) {
            while (nums[i] >= 1 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                const int j = nums[i] - 1;
                swap(nums[i], nums[j]);
            }
        }
        for (int i = 0; i < n; ++i) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        return n + 1;
    }
};

File: Hash_Table/problem41/problem41.py
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        for i in range(n):
            while (x := nums[i]) >= 1 and x <= n and nums[x - 1] != x:
                nums[i], nums[x - 1] = nums[x - 1], x
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
        return n + 1
